src/Core/Repository/AbstractMultiShopObjectModelRepository.php

```diff
@@ -0,0 +1,176 @@
+    {
+        $objectModel = $this->fetchObjectModel($id, $objectModelClass, $exceptionClass, $shopId->getValue());
+
+        // The object is fetched before checking the association, so that the NotFoundException has the priority over the NoAssociationException
+        $this->checkShopAssociation($id, $objectModelClass, $shopId);
+
+        // Force id_shop_list right away so that DB modification use the appropriate shop and not the one from context
+        $objectModel->id_shop_list = [$shopId->getValue()];
+
+        return $objectModel;
+    }
+
+    /**
+     * @param ObjectModel $objectModel
+     * @param string $exceptionClass
+     * @param int $errorCode
+     *
+     * @return int
+     */
+    protected function addObjectModelToShop(ObjectModel $objectModel, int $shopId, string $exceptionClass, int $errorCode = 0): int
+    {
+        // Force internal shop list which is used as an override of the one from Context when generating the SQL queries
+        // this way we can control exactly which shop is updated
+        $objectModel->id_shop_list = [$shopId];
+
+        return $this->addObjectModel($objectModel, $exceptionClass, $errorCode);
+    }
+
+    /**
+     * @param ObjectModel $objectModel
+     * @param array $shopIds
+     * @param string $exceptionClass
+     * @param int $errorCode
+     *
+     * @throws CoreException
+     */
+    protected function updateObjectModelForShops(
+        ObjectModel $objectModel,
+        array $shopIds,
+        string $exceptionClass,
+        int $errorCode = 0
+    ): void {
+        // Force internal shop list which is used as an override of the one from Context when generating the SQL queries
+        // this way we can control exactly which shop is updated
+        $objectModel->id_shop_list = $shopIds;
+
+        $this->updateObjectModel($objectModel, $exceptionClass, $errorCode);
+    }
+
+    /**
+     * @param ObjectModel $objectModel
+     * @param array $propertiesToUpdate
+     * @param array $shopIds
+     * @param string $exceptionClass
+     * @param int $errorCode
+     *
+     * @throws CoreException
+     */
+    protected function partiallyUpdateObjectModelForShops(
+        ObjectModel $objectModel,
+        array $propertiesToUpdate,
+        array $shopIds,
+        string $exceptionClass,
+        int $errorCode = 0
+    ): void {
+        $objectModel->setFieldsToUpdate($this->formatPropertiesToUpdate($propertiesToUpdate));
+        $this->updateObjectModelForShops($objectModel, $shopIds, $exceptionClass, $errorCode);
+    }
+
+    /**
+     * @param int $id
+     * @param string $objectModelClassName
+     * @param ShopId $shopId
+     *
+     * @return bool
+     */
+    protected function hasShopAssociation(int $id, string $objectModelClassName, ShopId $shopId): bool
+    {
+        $modelDefinition = $objectModelClassName::$definition;
+        $objectTable = $modelDefinition['table'];
+        $primaryColumn = $modelDefinition['primary'];
```

Comment:
Should we create some helper functions `getPrimaryColumn()` on ObjectModel to make it reusable? we could force the `bqSQL()` inside

Reply:
I can add a static method in ObjectModel for that indeed However I don't think the `bqSQL` must be included inside, in case the method is used for another purpose than creating an SQL request

`bqSQL` should only be used when build the request

Reply:
Actually, it's counterproductive, since I also need to get the table from the definition it's actually more efficient to get the whole definition and pick the necessary fields

I could create two static methods (one for table and one for primary column), but it seems like it becomes a bit overkill